### PR TITLE
fix: uncorrect exp spending on unlocking research

### PR DESCRIPTION
**Pull request by kingdom84521**
## Description

The amount of experience spending on unlocking researches is incorrect, since this process is done by Player.setLevel only.

## Proposed changes

1. Add some formula utils for calculating the proper value for unlocking
2. Refactor the original setLevel to new utils

## Checklist

- [v] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [v] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.19.*).
- [v] I followed the existing code standards and didn't mess up the formatting.
- [v] I did my best to add documentation to any public classes or methods I added.
- [v] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import javax.annotation.Nonnull;
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.utils.ExperienceUtils;
```
Comment (ybw0014) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java:8]: import order

**Review by ybw0014 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java (hunk 1 above)
Comment (kingdom84521) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java]: fix

**Comment by J3fftw1** [comment]
I feel like this should be a config option this has been this way for many years it would be weird to suddenly change this. Some addons would now also be easier to be unlocked for example I have uu matter set to 100 lvls taking in count that it’s actually levels

**Comment by kingdom84521** [comment]
That's a point, I'll try to add a option for that.

**Comment by Sefiraat** [comment]
The config option should be under the `research` heading. And I'm not a big fan of the name? `level-cost-as-exp`? Not sure that's much better xD

**Comment by kingdom84521** [comment]
Actually I put it in research at first, but looks like other options in research seems to be like some options that can alter by player on their own? That get me feel a little bit weird, so I move it to option at last.

Also, I'm a pretty bad namer, so if anyone thought a better name, feel free to tell me, I can fix it right away.

**Comment by SchnTgaiSpock** [comment]
`use-legacy-xp-spending`?

**Comment by kingdom84521** [comment]
hmm, I will prefer "vanilla" than "legacy", cause vanilla minecraft is still using this kind of formula on enchanting.

**Comment by JustAHuman-xD** [comment]
I agree with schnn I think it should be legacy, we are talking about slimefun, not minecraft

**Review by Sefiraat — CHANGES_REQUESTED** [review]
Not legacy, that implies we are changing the default which we are not. 

Go with what I have put for the name for now but let’s hope something better comes up. And it still belongs in research category regardless.

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package io.github.thebusybiscuit.slimefun4.utils;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class TestExperienceUtils {
```
Comment (TheBusyBiscuit) [inline src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java:7]:
```suggestion
class TestExperienceUtils {
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    void testConvertLevelToXp() {
+        // case: level >= 32
+        Assertions.assertEquals(5345F, ExperienceUtils.convertLevelToFloatExp(50));
+
+        // case: level >= 17 && level <= 31
+        Assertions.assertEquals(910F, ExperienceUtils.convertLevelToFloatExp(25));
+
+        // case: level <= 16
+        Assertions.assertEquals(160F, ExperienceUtils.convertLevelToFloatExp(10));
+    }
```
Comment (TheBusyBiscuit) [inline src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java:19]: Make sure to also always test edge cases like -1, 0, MAX_VALUE

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if (exp >= 1508) {
+            return (float) (325 + Math.sqrt(72 * exp - 54215)) / 18;
+        } else if (exp >= 353) {
+            return (float) (81 + Math.sqrt(40 * exp - 7839)) / 10;
+        } else {
+            return (float) Math.sqrt(exp + 9) - 3;
+        }
+    }
+
+    public static Number[] convertExpToLevelAndProgress(float exp) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:60]: I think a proper record/struct or small class would be much better than a `Number` array.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Some helper methods for dealing with experience converting.
+ *
+ * @author kingdom84521
+ *
+ * @see io.github.thebusybiscuit.slimefun4.core.guide.SlimefunGuideImplementation
+ * @see <a href="https://minecraft.fandom.com/wiki/Experience"></a>
+ *
+ */
+public final class ExperienceUtils {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:15]: Please also add javadocs to all the public methods in this class, it's a utility class after all.

File: src/main/resources/config.yml [hunk 6]
```diff
@@ -29,6 +29,7 @@ researches:
   free-in-creative-mode: true
   enable-fireworks: true
   disable-learning-animation: false
+  convert-level-cost-to-exp: false
```
Comment (TheBusyBiscuit) [inline src/main/resources/config.yml:32]: Please rename this.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java [hunk 7]
```diff
@@ -70,7 +71,15 @@ default void unlockItem(Player p, SlimefunItem sfitem, Consumer<Player> callback
         if (p.getGameMode() == GameMode.CREATIVE && Slimefun.getRegistry().isFreeCreativeResearchingEnabled()) {
             research.unlock(p, true, callback);
         } else {
-            p.setLevel(p.getLevel() - research.getCost());
+            if (Slimefun.getCfg().getBoolean("researches.convert-level-cost-to-exp")) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java:74]: We shouldn't read this from the config everytime a research is unlocked for performance reasons. Look into how it's handled with over config settings like the learning animation for example.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Not sure what to think of this.

The XP behaviour has been consistently unchanged for almost TEN years. Changing it now will just result in a ton of bug tickets for weeks to come. Good that this is disabled by default, so if there really is a demand for this setting then I am willing to merge it

**Comment by Sefiraat** [comment]
From my experience on a few servers now it is generally considered to be unintuitive to players coming into slimefun blind but once known it actually becomes a mechanic that rewards forward thinking and/or the use of mechanics offered in addons/3rd party plugins to make the most of xp. 

While I prefer the current system in the main, I'm all for the change so long as it's default-off like this.

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java (hunk 3 above)
Comment (kingdom84521) [reply src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java:19]: I...don't really know how to produce a test case for MAX_VALUE, but I add the other ones' case

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java (hunk 2 above)
Comment (kingdom84521) [reply src/test/java/io/github/thebusybiscuit/slimefun4/utils/TestExperienceUtils.java:7]: fix

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java (hunk 4 above)
Comment (kingdom84521) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:60]: fix

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java (hunk 5 above)
Comment (kingdom84521) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:15]: fix

File: src/main/resources/config.yml (hunk 6 above)
Comment (kingdom84521) [reply src/main/resources/config.yml:32]: fix

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java (hunk 7 above)
Comment (kingdom84521) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/guide/SlimefunGuideImplementation.java:74]: fix

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    public static class LevelAndProgress {
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:16]: We shouldnt use inner classes

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private static float getExpToNextLevel(int currentLevel) {
+        if (currentLevel > 30) {
+            return (float) 9 * currentLevel - 158;
+        } else if (currentLevel > 15) {
+            return (float) 5 * currentLevel - 38;
+        } else {
+            return (float) 2 * currentLevel + 7;
+        }
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/ExperienceUtils.java:54]: not a fan of these methods. Isnt there a better way to do this?
I dont think the calculation is correct to be honest.

**Review by J3fftw1 — CHANGES_REQUESTED** [review]
Sorry for the late review.

**Comment by kingdom84521** [comment]
Sorry for the lack of focusing on this PR, my tasks were finally done and dusted, it's been a nightmare last year :(
